tsinfer/formats.py

```diff
@@ -2007,30 +2064,69 @@ def num_alleles(self, sites=None):
     def variants(self, sites=None):
         """
         Returns an iterator over the Variant objects. This is equivalent to
-        the TreeSequence.variants iterator.
+        the TreeSequence.variants iterator. The alleles are guaranteed to be
+        returned in an order such that the ancestral state (if know) is the
+        first item in the list (and hence a ``0`` in the genotypes array) corresponds
+        to a sample with the ancestral state. This may mean that the indexes in the
+        genotypes array are recoded compared to the array passed in to :meth:`add_site`.
+
+        :param array sites: A numpy array of ascending site ids for which to return
+            data. If None (default) return all sites.
+        :return: An interator over the variants in the sample data file.
+        :rtype: iter(Variant)
+
         """
         all_genotypes = chunk_iterator(self.sites_genotypes, indexes=sites)
-        for genotypes, site in zip(all_genotypes, self.sites(ids=sites)):
-            yield Variant(site=site, alleles=site.alleles, genotypes=genotypes)
+        ancestral_allele_index = self.sites_ancestral_allele_index[:]
+        if sites is not None:
+            ancestral_allele_index = ancestral_allele_index[np.array(sites)]
+        for genos, site, aa_index in zip(
+            all_genotypes,
+            self.sites(ids=sites),
+            ancestral_allele_index,
+        ):
+            if aa_index > 0:
+                # re-map the genotypes
+                geno_map = np.arange(len(site.alleles) + 1)
+                geno_map[MISSING_DATA] = MISSING_DATA
+                geno_map[aa_index] = 0
+                geno_map[0:aa_index] += 1
+                genos = geno_map[genos]
+            yield Variant(site=site, alleles=site.alleles, genotypes=genos)
 
     def __all_haplotypes(self, sites=None):
         # We iterate over chunks vertically here, and it's not worth complicating
         # the chunk iterator to handle this.
+        aa_indexes = self.sites_ancestral_allele_index[:]
+        # If ancestral allele is missing, keep the order unchanged (aa_index of zero)
+        aa_indexes[aa_indexes == MISSING_DATA] = 0
         chunk_size = self.sites_genotypes.chunks[1]
         for j in range(self.num_samples):
             if j % chunk_size == 0:
                 chunk = self.sites_genotypes[:, j : j + chunk_size].T
             a = chunk[j % chunk_size]
+            # Remap the genotypes at all sites, depending on the ancestral_allele index
```

Comment:
Nice - not totally following what's going on, but can't see this being a bottleneck.